src/benchmarks/micro/corefx/System.Collections/CopyAndRemove.cs

```diff
@@ -0,0 +1,173 @@
+
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Extensions;
+using MicroBenchmarks;
+
+namespace System.Collections
+{
+    [BenchmarkCategory(Categories.CoreFX, Categories.Collections, Categories.GenericCollections)]
+    [GenericTypeArguments(typeof(int))] // value type
+    [GenericTypeArguments(typeof(string))] // reference type
+    [GenericTypeArguments(typeof(CustomValue))] // custom value type (more overhead)
+    [GenericTypeArguments(typeof(CustomObject))] // custom reference type (less overhead)
+    public class CopyAndRemove<T>
+    {
+        private T[] _items;
+        
+        private List<T> _list;
+        private LinkedList<T> _linkedList;
+        private HashSet<T> _hashSet;
+        private SortedSet<T> _sortedSet;
+        private Dictionary<T, T> _dictionary;
+        private SortedList<T, T> _sortedList;
+        private SortedDictionary<T, T> _sortedDictionary;
+
+        [Params(Utils.DefaultCollectionSize)]
+        public int Size;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            T[] twoParts = ValuesGenerator.ArrayOfUniqueValues<T>(Size * 2);
+            T[] part1 = twoParts.Take(Size).ToArray();
+            T[] part2 = twoParts.Skip(Size).ToArray();
+
+            _items = part1;
+            _list = new List<T>(_items);
+            _linkedList = new LinkedList<T>(_items);
+            _hashSet = new HashSet<T>(_items);
+            _sortedSet = new SortedSet<T>(_items);
+            _dictionary = new Dictionary<T, T>();
+            _sortedList = new SortedList<T, T>();
+            _sortedDictionary = new SortedDictionary<T, T>();
+
+            foreach (var value in _items)
+            {
+                _dictionary.Add(value, value);
+                _sortedList.Add(value, value);
+                _sortedDictionary.Add(value, value);
+            }
+
+            Array.Sort(part2, _items);
+        }
+
+        [Benchmark]
+        public bool List()
+        {
+            bool result = true;
+            T[] items = _items;
+            List<T> copy = new List<T>(_list);
+
+            for (int i = 0; i < items.Length; i++)
+                result &= copy.Remove(items[i]);
+
+            if (!result)
+                throw new InvalidOperationException("All items must be in the collection.");
```

Comment:
We should not be throwing inside the timing loop. We don't want to be doing any unnecessary work inside of the timing loop, and we shouldn't be doing correctness checks in performance tests. I know that this is a pattern in other parts of the benchmarks, but we should be working to remove those as well.

Reply:
May I ask why we shouldn't do that? The check is per iteration instead of per inner-loop of each `Remove` call so it shouldn't affect the benchmark result. The code is to ensure that the data prepared is what we want, since like `RemoveTrue` and `RemoveFalse` are two distinct cases that we want to test, so IMHO the code should be able catch the issue here.